In [12]:
import graphlab

In [13]:
sales = graphlab.SFrame('kc_house_data.gl/')

In [14]:
train_data,test_data = sales.random_split(.8,seed=0)

### Simple Linear Regression

In [15]:
def simple_linear_regression(input_feature, output):
    input_sum = input_feature.sum()
    output_sum = output.sum()
    
    n = input_feature.size()   
    output_mean = output_sum / n
    input_mean = input_sum / n
    
    input_multiply_output_mean = (input_feature * output).sum() / n
    
    input_squared = input_feature * input_feature
    input_squared_mean = input_squared.sum() / n
    
    numerator = input_multiply_output_mean - (output_mean * input_mean)
    denumerator = input_squared_mean - (input_mean * input_mean)
    
    slope = numerator / denumerator
    intercept = output_mean - (slope * input_mean)
    return (intercept, slope)


Test simple_linear_regression

In [16]:

test_feature = graphlab.SArray(range(5))
test_output = graphlab.SArray(1 + 1*test_feature)
(test_intercept, test_slope) =  simple_linear_regression(test_feature, test_output)
print "Intercept: " + str(test_intercept)
print "Slope: " + str(test_slope)


Intercept: 1
Slope: 1


In [17]:
def get_regression_predictions(input_feature, intercept, slope):
    predicted_output = (input_feature * slope) + intercept
    return(predicted_output)

#### Q6. Quiz Question: Using your Slope and Intercept from (4), What is the predicted price for a house with 2650 sqft?

In [18]:
q6_input_feature = graphlab.SArray([2650])
input_feature = train_data['sqft_living']
output = train_data['price']

(intercept, slope) = simple_linear_regression(input_feature, output)
q6_answer = get_regression_predictions(q6_input_feature, intercept, slope)
q6_answer

dtype: float
Rows: 1
[700074.8456294583]

### RSS

In [31]:
def get_residual_sum_of_squares(input_feature, output, intercept, slope):
    # First get the predictions
    predictions = get_regression_predictions(input_feature, intercept, slope)
    
    # then compute the residuals (since we are squaring it doesn't matter which order you subtract)
    residuals = predictions - output
    residuals_squared = residuals * residuals
    
    # square the residuals and add them up
    RSS = residuals_squared.sum()
    
    return(RSS)

#### 8. Quiz Question: According to this function and the slope and intercept from (4) What is the RSS for the simple linear regression using squarefeet to predict prices on TRAINING data?

In [44]:
print 'Test should be zero: ' + str(get_residual_sum_of_squares(test_feature, test_output, test_intercept, test_slope))

print 'RSS is ' + str(get_residual_sum_of_squares(input_feature, output, intercept, slope))
print 'RSS is %.2f' % get_residual_sum_of_squares(input_feature, output, intercept, slope)

Test should be zero: 0
RSS is 1.20191835632e+15
RSS is 1201918356321968.00


### Inverse linear function and get sq.ft based on price

In [50]:
def inverse_regression_predictions(output, intercept, slope):
    estimated_input = (output - intercept)/slope
    return(estimated_input)

#### 10. Quiz Question: According to this function and the regression slope and intercept from (3) what is the estimated square-feet for a house costing $800,000?

In [56]:
house_price = 800000
print 'Price of $%d should be %.2f sq.ft' % (house_price, inverse_regression_predictions(house_price, intercept, slope))

Price of $800000 should be 3004.40 sq.ft


### Using bedrooms as input

In [68]:
bedroom_feature = input_feature = train_data['bedrooms']
(bedroom_intercept, bedroom_slope) = simple_linear_regression(bedroom_feature, output)   
print 'Bedroom intercept: %.2f, bedroom slope %.2f' % (bedroom_intercept, bedroom_slope)

print 'Bedroom RSS is %.2f' % get_residual_sum_of_squares(bedroom_feature, output, bedroom_intercept, bedroom_slope)

Bedroom intercept: 109473.18, bedroom slope 127588.95
Bedroom RSS is 2143244494226585.25


#### 13. Quiz Question: Which model (square feet or bedrooms) has lowest RSS on TEST data? Think about why this might be the case.

Bedroom RSS is : 2,143,244,494,226,585.25


Sq.ft RSS is   : 1,201,918,356,321,968.00 